backend/src/utils/proxy.ts

```diff
@@ -74,56 +77,93 @@ export const proxyService =
+                `Proxy could not find address annotation on k8s service ${serviceName} in namespace ${namespace}, falling back to internal address. Annotation expected: ${service.addressAnnotation}`,
+              );
+            } catch (e) {
+              fastify.log.error(
+                e,
+                `Proxy failed to read k8s service ${serviceName} in namespace ${namespace}.`,
+              );
+              return null;
+            }
+          }
+          // For services configured for internal addresses (no annotation), construct the URL
+          // or if annotation is expected but missing, fall back to internal address for compatibility
+          return `${scheme}://${serviceName}.${namespace}.svc.cluster.local:${service.internalPort}`;
+        };
 
-          // assign the `upstream` param so we can dynamically set the upstream URL for http-proxy
+        const doServiceRequest = async (resource?: K) => {
+          const upstream = await getServiceAddress(serviceName, resource);
+          if (!upstream) {
+            done(notFoundError('Service', serviceName, undefined, 'service unavailable'));
+            return;
+          }
+          // Assign the `upstream` param so we can dynamically set the upstream URL for http-proxy
           setParam(request, 'upstream', upstream);
-
+          if (tls) {
+            const requestOptions = await getDirectCallOptions(fastify, request, '');
+            const token = getAccessToken(requestOptions);
+            request.headers.authorization = `Bearer ${token}`;
+          }
           fastify.log.info(`Proxy ${request.method} request ${request.url} to ${upstream}`);
           done();
         };
 
-        if (model) {
-          const kc = fastify.kube.config;
-          const cluster = kc.getCurrentCluster();
-
-          // retreive the gating resource by name and namespace
-          passThroughResource<K>(fastify, request, {
-            url: `${cluster.server}/apis/${model.apiGroup}/${model.apiVersion}/namespaces/${namespace}/${model.plural}/${name}`,
-            method: 'GET',
-          })
-            .then((resource) => {
-              return getDirectCallOptions(fastify, request, request.url).then((requestOptions) => {
-                if (isK8sStatus(resource)) {
-                  done(notFoundError(model.kind, name));
-                } else if (!statusCheck || statusCheck(resource)) {
-                  if (tls) {
-                    const token = getAccessToken(requestOptions);
-                    request.headers.authorization = `Bearer ${token}`;
-                  }
-
-                  doServiceRequest(resource);
-                } else {
-                  done(notFoundError(model.kind, name, undefined, 'service unavailable'));
-                }
-              });
-            })
-            .catch((e) => {
-              done(notFoundError(model.kind, name, e));
+        const doServiceRequestWithGatingResource = async () => {
```

Comment:
Can you add some comments to understand the difference between with/without gating resources? it took me a little bit to understand and every other step is documented.